clusterloader2/docs/DEVELOPING_MEASUREMENT.md

```diff
@@ -0,0 +1,579 @@
+Our measurement will have two actions `start` and `gather` so also want to know when our measurement is in running state.
+We will need also channel for stopping informers.
+```golang
+type maxRunningPodsMeasurement struct {
+	maxRunningPods     int
+	currentRunningPods int
+	stopCh             chan struct{}
+	isRunning          bool
+}
+
+func createMaxRunningPodsMeasurement() measurement.Measurement {
+	return &maxRunningPodsMeasurement{}
+}
+```
+
+Once we have it, we want to register our new measurement so it can be used in test config:
+```golang
+func init() {
+	if err := measurement.Register(maxRunningPodsMeasurementName, createMaxRunningPodsMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", maxRunningPodsMeasurementName, err)
+	}
+}
+```
+
+Next, let's implement `String` method:
+```golang
+func (*maxRunningPodsMeasurement) String() string {
+	return maxRunningPodsMeasurementName
+}
+```
+Next, we need to implement `Execute` method. We want to have two actions - `start` and `gather`:
+```golang
+func (s *maxRunningPodsMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "start":
+		return nil, s.start(config.ClusterFramework.GetClientSets().GetClient())
+	case "gather":
+		return s.gather()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+```
+
+Now, let's focus on `start` method. First, we want to check if measurement is not already running. In this case we want to return error.
+If measurement is not running then we want to start informer that will get information about all pods within cluster.
+
+```golang
+func (m *maxRunningPodsMeasurement) start(c clientset.Interface) error {
+	if m.isRunning {
+		return fmt.Errorf("%s: measurement already running", m)
+	}
+	klog.V(2).Infof("%s: starting max pod measurement...", m)
+	m.isRunning = true
+	m.stopCh = make(chan struct{})
+	i := informer.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return c.CoreV1().Pods("").List(context.TODO(), options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return c.CoreV1().Pods("").Watch(context.TODO(), options)
+			},
+		},
+		m.checkPod,
+	)
+	return informer.StartAndSync(i, m.stopCh, informerSyncTimeout)
+}
+
+```
+
+Next step is implementing `checkPod` method. This method will be responsible for counting pods in Running state.
+As arguments, we get old and new version of object.
+There are two special cases.
+First case when old version of object is nil - it means that object is added.
+Second case when new version of object is nil - it means that object was deleted.
+
+```golang
+func MaxInt(x, y int) int {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+func (m *maxRunningPodsMeasurement) checkPod(oldObj, newObj interface{}) {
```

Comment:
Discuss why we don't need locking?

[or rather why we're ignoring it in all measurements...]

Reply:
Added lock